extension/js/common/core/crypto/pgp/openpgp-key.ts

```diff
@@ -0,0 +1,417 @@
+    }
+    if (encryptedPacketCount) {
+      throw new Error(`Cannot encrypt a key that has ${encryptedPacketCount} of ${secretPackets.length} private packets still encrypted`);
+    }
+    await prv.encrypt(passphrase);
+    if (!prv.isFullyEncrypted()) {
+      throw new Error('Expected key to be fully encrypted after prv.encrypt');
+    }
+    await OpenPGPKey.wrap(prv, key);
+  }
+
+  public static decrypt = async (message: OpenPGP.message.Message, privateKeys: Key[], passwords?: string[]) => {
+    return await message.decrypt(privateKeys.map(key => OpenPGPKey.unwrap(key)), passwords, undefined, false);
+  }
+
+  public static encrypt: PgpMsgMethod.Encrypt = async ({ pubkeys, signingPrv, pwd, data, filename, armor, date }) => {
+    const message = opgp.message.fromBinary(data, filename, date);
+    const options: OpenPGP.EncryptOptions = { armor, message, date };
+    let usedChallenge = false;
+    if (pubkeys) {
+      options.publicKeys = [];
+      for (const pubkey of pubkeys) {
+        const { keys: publicKeys } = await opgp.key.readArmored(KeyUtil.armor(pubkey));
+        options.publicKeys.push(...publicKeys);
+        // TODO: Investigate why unwrapping doesn't work - probably the object
+        // came from the background page so it wasn't properly deserialized
+        // options.publicKeys.push(OpenPGPKey.unwrap(pubkey));
+      }
+    }
+    if (pwd) {
+      options.passwords = [await PgpHash.challengeAnswer(pwd)];
+      usedChallenge = true;
+    }
+    if (!pubkeys && !usedChallenge) {
+      throw new Error('no-pubkeys-no-challenge');
+    }
+    if (signingPrv) {
+      const openPgpPrv = OpenPGPKey.unwrap(signingPrv);
+      if (typeof openPgpPrv.isPrivate !== 'undefined' && openPgpPrv.isPrivate()) { // tslint:disable-line:no-unbound-method - only testing if exists
+        options.privateKeys = [openPgpPrv];
+      }
+    }
+    const result = await opgp.encrypt(options);
+    if (typeof result.data === 'string') {
+      return { data: Buf.fromUtfStr(result.data), signature: result.signature, type: 'openpgp' };
+    } else {
+      return result as unknown as OpenPGP.EncryptBinaryResult;
+    }
+  }
+
+  public static isWithoutSelfCertifications = async (key: Key) => {
+    const k = OpenPGPKey.unwrap(key);
+    return await Catch.doesReject(k.verifyPrimaryKey(), ['No self-certifications']);
+  }
+
+  public static reformatKey = async (privateKey: Key, passphrase: string, userIds: { email: string | undefined; name: string }[], expireSeconds: number) => {
+    const origPrv = OpenPGPKey.unwrap(privateKey);
+    const keyPair = await opgp.reformatKey({ privateKey: origPrv, passphrase, userIds, keyExpirationTime: expireSeconds });
+    return await OpenPGPKey.wrap(keyPair.key, {} as Key);
+  }
+
+  // TODO: should be private, will change when readMany is rewritten
+  public static wrap = async (pubkey: OpenPGP.key.Key, pkey: Key, raw?: string): Promise<Key> => {
```

Comment:
Maybe we could find another name for this, but maybe not. I'm thinking `toKeyObj` which may be a worse name.

Reply:
:thinking: or `toPubkey` or however will that interface be called. I meant this to be used internally but as the comment explains this is still not yet possible.